test/summary_truncate.c

```diff
@@ -0,0 +1,215 @@
+}
+
+void
+it_omits_ON_CONFLICT_target_list(TestState * test_state)
+{
+	TEST_INIT();
+	char	   *query = "INSERT INTO y(a) VALUES(1) ON CONFLICT DO UPDATE SET a = 123456789";
+	Summary		result = summary(query, 0, 65);
+
+	TEST_ASSERT_STR_EQUAL(result.truncated_query, "INSERT INTO y (a) VALUES (1) ON CONFLICT DO UPDATE SET ... = ...");
+}
+
+void
+it_omits_ON_CONFLICT_target_list_2(TestState * test_state)
+{
+	TEST_INIT();
+	char	   *query = "\
+        INSERT INTO foo (a, b, c, d) VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16,$17,$18,$19,$20,$21,$22,$23,$24,$25,$26,$27,$28,$29) \
+        ON CONFLICT (id) \
+        DO UPDATE SET (a, b, c, d) = (excluded.a,excluded.b,excluded.c,case when foo.d = excluded.d then excluded.d end) \
+    ";
+	Summary		result = summary(query, 0, 100);
+
+	TEST_ASSERT_STR_EQUAL(result.truncated_query, "INSERT INTO foo (a, b, c, d) VALUES (...) ON CONFLICT (id) DO UPDATE SET ... = ...");;
+}
+
+void
+it_handles_GRANT(TestState * test_state)
+{
+	TEST_INIT();
+	char	   *query = "GRANT SELECT (abc, def, ghj) ON TABLE t1 TO r1";
+	Summary		result = summary(query, 0, 35);
+
+	TEST_ASSERT_STR_EQUAL(result.truncated_query, "GRANT select (abc, def, ghj) ON ...");
+}
+
+void
+it_does_not_segfault_on_target_list_from_CTE_already_removed_from_possible_truncations(TestState * test_state)
+{
+	TEST_INIT();
+	char	   *query = "\
+        WITH activity AS ( \
+            SELECT pid, COALESCE(a.usename, '') AS usename \
+            FROM pganalyze.get_stat_activity() a \
+        ) \
+        SELECT \
+        FROM pganalyze.get_stat_progress_vacuum() v \
+        JOIN activity a USING (pid) \
+    ";
+	Summary		result = summary(query, 0, 100);
+
+	TEST_ASSERT_STR_EQUAL(result.truncated_query, "WITH activity AS (...) SELECT FROM pganalyze.get_stat_progress_vacuum() v JOIN activity a USING (...");;
+}
+
+/*  If we truncate the index expression in the future this would remove (lower(d) || upper(d)) first */
+void
+it_handles_CREATE_INDEX(TestState * test_state)
+{
+	TEST_INIT();
+	char	   *query = "CREATE INDEX testidx ON test USING btree ((lower(d) || upper(d)), a, (b+c))";
+	Summary		result = summary(query, 0, 60);
+
+	TEST_ASSERT_STR_EQUAL(result.truncated_query, "CREATE INDEX testidx ON test USING btree ((lower(d) || up...");;
+}
+
+void
+char_truncate_works(TestState * test_state)
+{
+	char	   *query = "WITH \"原チコ氏にはす腹腹腹腹腹腹腹腹腹腹腹\" AS (SELECT) SELECT w";
+	Summary		result = summary(query, 0, 21);
+	char	   *output = "WITH \"原チコ氏にはす腹腹腹腹腹...";
+
+	TEST_ASSERT_STR_EQUAL(result.truncated_query, output);;
+}
+
+/*
+#[should_panic(
+    expected = "byte index 22 is not a char boundary; it is inside 'は' (bytes 21..24) of `WITH \"原チコ氏にはす腹腹腹腹腹腹腹腹腹腹腹\" AS (SELECT) SELECT w`"
+)]
+*/
+/*
+void byte_truncate_fails(TestState* test_state) {
+    char* query = "WITH \"原チコ氏にはす腹腹腹腹腹腹腹腹腹腹腹\" AS (SELECT) SELECT w".to_string();
+    query[0..=21].to_string();
+}
+*/
```

Comment:
I'm not sure how to re-add this. I don't think the C test infrastructure supports an equivalent right now.

Reply:
Yes it was added at the same time as the test above it, documenting string truncation behavior in Rust: https://github.com/pganalyze/pg_query.rs/pull/7

Reply:
Okay, so since we're truncating to a certain number of characters (not bytes) we don't have that problem. Thanks for confirming that. ^^